ton_client/src/processing/send_message.rs

```diff
@@ -128,32 +126,91 @@ impl SendingMessage {
+                futures.push(Box::pin(async move {
+                    message.send_to_endpoint(context, &address).await
+                }));
+            }
+            for result in futures::future::join_all(futures).await {
+                match result {
+                    Ok(_) => {
+                        if let Some(callback) = &callback {
+                            callback(ProcessingEvent::DidSend {
+                                shard_block_id: shard_block_id.to_string(),
+                                message_id: self.id.clone(),
+                                message: self.serialized.clone(),
+                            })
+                            .await
+                        }
+                        return Ok(());
+                    }
+                    Err(_) => {
+                        // If one of the endpoints responds with clock out of sync error
+                        // we leave this error as a final error.
+                        // This is required by out of sync unit test.
+                        let last_result_is_out_of_sync = if let Some(Err(err)) = &last_result {
+                            err.code == crate::net::ErrorCode::ClockOutOfSync as u32
+                        } else {
+                            false
+                        };
+                        if !last_result_is_out_of_sync {
+                            last_result = Some(result);
+                        }
+                    }
+                }
+            }
+        }
 
+        let result =
+            last_result.unwrap_or_else(|| Err(Error::block_not_found("no endpoints".to_string())));
         if let Some(callback) = &callback {
-            callback(match send_error {
-                None => ProcessingEvent::DidSend {
-                    shard_block_id: shard_block_id.clone(),
+            if let Err(err) = &result {
+                callback(ProcessingEvent::SendFailed {
+                    shard_block_id: shard_block_id.to_string(),
                     message_id: self.id.clone(),
                     message: self.serialized.clone(),
-                },
-                Some(error) => ProcessingEvent::SendFailed {
-                    shard_block_id: shard_block_id.clone(),
-                    message_id: self.id.clone(),
-                    message: self.serialized.clone(),
-                    error: Error::send_message_failed(error, &self.id, &shard_block_id),
-                },
-            })
-            .await;
+                    error: Error::send_message_failed(err, &self.id, shard_block_id),
+                })
+                .await;
+            }
         }
-        Ok(ResultOfSendMessage { shard_block_id })
+        result
+    }
+
+    async fn send_to_endpoint(
+        &self,
+        context: Arc<ClientContext>,
+        endpoint_address: &str,
+    ) -> ClientResult<()> {
+        let endpoint = Endpoint::resolve(context.env.clone(), endpoint_address).await?;
```

Comment:
Each send consist of 3 requests to DApp server: version request, time request and send intself. Version and time requests could be merged to reduce send time